torch/distributed/rpc/api.py

```diff
@@ -104,17 +160,21 @@ def shutdown(graceful=True):
         >>> rpc.shutdown()
     """
     global _agent
-    if _agent:
-        if graceful:
-            _wait_all_workers()
-        _destroy_rref_context(_ignore_rref_leak)
-        _agent.shutdown()
-        # clean up python rpc handler in shutdown(), see comments in
-        # PythonRpcHandler::cleanup(), call it in python API because the
-        # cleanup() function has python dependency, it assumes python
-        # interpreter exists
-        _cleanup_python_rpc_handler()
-        _agent = None
+
+    if _agent is None:
+        return
+
+    if graceful:
+        _wait_all_workers()
+    _destroy_rref_context(_ignore_rref_leak)
+    _agent.shutdown()
```

Comment:
So, I guess, with the current implementation, we won't wait for the very last RPC sent by the leader worker to be processed, since `_agent.shutdown()` does not guarantee that it will wait for outstanding RPCs?

Reply:
@rohan-varma 

I noticed in `ProcessGroupAgent::shutdown()` we do wait for the thread pool to finish. The problem is that we can't guarantee the message is in the thread pool queue before calling `shutdown()`.

I do noticed there is flaky test cases that timed out.

I have to add another atomic flag in C++ `RpcAgent`, and expose it to the Python world, once we are in the `_wait_all_workers` we set this flag. `RpcAgent::shutdown()` wait this flags to be flipped before really shutting down.

Reply:
@rohan-varma 
@satgera 

I added more loggings. 
I sew the`SendWork`s of the last a few requests for allowing followers to proceed had been run by the thread pool. 

Need to confirm where the request messages got lost. I think now it's in GLOO buffer or below.

Reply:
@rohan-varma For now, we are using a sleep.